rmw_fastrtps_cpp/src/functions.cpp

```diff
@@ -648,6 +648,7 @@ class ClientListener : public SubscriberListener
 
   bool hasData()
   {
+    std::lock_guard<std::mutex> lock(internalMutex_);
     return !list.empty();
```

Comment:
You could store this state in an atomic boolean variable just before any calls to notify `conditionVariable_` after `list` is modified. That way you would not need a lock here.